oeps/oep-0031-arch-i18n.rst

```diff
@@ -0,0 +1,212 @@
+
+***********************
+Translations Management
+***********************
+
+Message Ids
+===========
+
+Message ids, or message keys, uniquely identify a string to be internationalized. They should consist of one or more words to namespace the message, plus one or more words to identify the message. All words will be lowercase and separated by periods. Here are examples of message ids used in a header and in a login modal, respectively:
+
+  | ``header.search``
+  | ``login.no.such.password``
+  | ``login.privacy.policy``
+
+The part of the id used for namespacing is purely to reduce collisions and help human readability; it's not intended to be programmatically parsable. That said, a helpful convention is to make it the name of your application or repo. Note that if you end up with a duplicate message id in the same translation project in Transifex, and they have different content, one of them will "win" when you download your translated strings, leading to an out-of-place string in your application.
+
+Alternatives Considered
+-----------------------
+
+Previous i18n implementations have sometimes used the English text as the message id. While this saves developers a step in their work, it has the drawback that a change in the English copy invalidates all existing translations, even if they would not otherwise have needed to change. (For example, capitalizing a noun in English has no impact on the German translation, because nouns are capitalized in German anyway. Or an English change might be correcting a typo that was correct in the translations.) Do not use English copy as the message id.
+
+Cross-Site Scripting Vulnerability
+==================================
+
+Because translated strings often contain placeholders that will be filled in at render time, possibly using user input, localization is a potential opening for cross-site scripting (XSS) attacks. The react-intl functions exposed by frontend-i18n are XSS-safe.
+
+Mapping Server Codes To Messages
+================================
+
+When a server response can contain one of a set of enumerated values, it can be tempting to assemble your message id on the fly by concatenating a common prefix and the server-provided value:
+
+``const id = 'myapp.errors.${errorCode}';``
+
+However, an unexpected value from the server can cause this technique to generate a message id that doesn't exist. If you choose to do this, it's best to whitelist your server-provided value against a list of expected responses.
+
+Missing Translations
+====================
+
+Missing translations can be a disorienting experience for a user. We strongly suggest a careful evaluation of translation coverage, or a clear indicator of beta status, before making a new language available to the public.
```

Comment:
Consider linking to the Beta Language feature as an example of how we are already doing this in the platform.

Reply:
Ugh.  So Beta Languages only manifests in the old account settings view... which means we've effectively blown away the feature as we permanently migrate to the account settings MFE.

Since edx.org has been running the account settings MFE in production for the last two years, Beta Languages seems like a candidate for DEPR rather than something we should include here.  Thoughts?

Reply:
Apparently I'm wrong, seeing your screenshot below.

Reply:
yeah. I think there's value in leaving these beta lang discussions open in case another reviewer can provide any more insight